src/keymgmt.c

```diff
@@ -2616,6 +2621,471 @@ const OSSL_DISPATCH p11prov_mldsa87_keymgmt_functions[] = {
+    if (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) {
+        const OSSL_PARAM *p;
+        p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PRIV_KEY);
+        if (!p) {
+            /* not really a private key */
+            class = CKO_PUBLIC_KEY;
+        }
+    }
+
+    rv = p11prov_obj_import_key(key, CKK_ML_KEM, class, param_set, params);
+    if (rv != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static int p11prov_mlkem_512_import(void *keydata, int selection,
+                                    const OSSL_PARAM params[])
+{
+    return p11prov_mlkem_import(keydata, selection, params, CKP_ML_KEM_512);
+}
+
+static int p11prov_mlkem_768_import(void *keydata, int selection,
+                                    const OSSL_PARAM params[])
+{
+    return p11prov_mlkem_import(keydata, selection, params, CKP_ML_KEM_768);
+}
+
+static int p11prov_mlkem_1024_import(void *keydata, int selection,
+                                     const OSSL_PARAM params[])
+{
+    return p11prov_mlkem_import(keydata, selection, params, CKP_ML_KEM_1024);
+}
+
+static int p11prov_mlkem_export(void *keydata, int selection,
+                                OSSL_CALLBACK *cb_fn, void *cb_arg)
+{
+    P11PROV_OBJ *key = (P11PROV_OBJ *)keydata;
+    P11PROV_CTX *ctx = p11prov_obj_get_prov_ctx(key);
+    CK_OBJECT_CLASS class = p11prov_obj_get_class(key);
+
+    P11PROV_debug("mlkem export %p, selection= %d", keydata, selection);
+
+    if (key == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    if (p11prov_ctx_allow_export(ctx) & DISALLOW_EXPORT_PUBLIC) {
+        return RET_OSSL_ERR;
+    }
+
+    /* if anything else is asked for we can't provide it, so be strict */
+    if ((class == CKO_PUBLIC_KEY) || (selection & ~(PUBLIC_PARAMS)) == 0) {
+        return p11prov_obj_export_public_key(key, CKK_ML_KEM, true, false,
+                                             cb_fn, cb_arg);
+    }
+
+    return RET_OSSL_ERR;
+}
+
+static const OSSL_PARAM *p11prov_mlkem_import_types(int selection)
+{
+    static const OSSL_PARAM p11prov_mlkem_imp_key_types[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PRIV_KEY, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PUB_KEY, NULL, 0),
+        OSSL_PARAM_END,
+    };
+    P11PROV_debug("mlkem import types");
+    if (selection & OSSL_KEYMGMT_SELECT_KEYPAIR) {
+        return p11prov_mlkem_imp_key_types;
+    }
+    return NULL;
+}
+
+static const OSSL_PARAM *p11prov_mlkem_export_types(int selection)
+{
+    static const OSSL_PARAM p11prov_mlkem_exp_key_types[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PUB_KEY, NULL, 0),
+        OSSL_PARAM_END,
+    };
+    P11PROV_debug("mlkem export types");
+    if (selection == OSSL_KEYMGMT_SELECT_PUBLIC_KEY) {
+        return p11prov_mlkem_exp_key_types;
+    }
+    return NULL;
+}
+
+/* See FIPS-203, 4. Parameter Sets */
+#define ML_KEM_512_MAX_SIZE 768
+#define ML_KEM_768_MAX_SIZE 1088
+#define ML_KEM_1024_MAX_SIZE 1568
```

Comment:
these match the `ML_KEM_*_CIPHERTEXT_BYTES` defines from the header file. Do we need to define them separately?